pkg/venafi/tpp/connector.go

```diff
@@ -1232,25 +1244,65 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
 	}
 
 	startTime := time.Now()
+	retrieveCount := 0
 	for {
-		var retrieveResponse *certificateRetrieveResponse
-		retrieveResponse, err = c.retrieveCertificateOnce(certReq)
+		retrieveCount++
+		statusCode, _, body, err := c.request("POST", urlResourceCertificateRetrieve, certReq)
 		if err != nil {
 			return nil, fmt.Errorf("unable to retrieve: %s", err)
 		}
-		if retrieveResponse.CertificateData != "" {
-			certificates, err = newPEMCollectionFromResponse(retrieveResponse.CertificateData, req.ChainOption)
+		var resp certificateRetrieveResponse
+		err = json.Unmarshal(body, &resp)
+		if err != nil {
+			return nil, fmt.Errorf("retrieve response body is unexpected: %s, HTTP status was %d and body was: %s", err, statusCode, body)
+		}
+		switch {
+		// After many tests, we found that the only way to know if the current
+		// certificate request is stuck due to an old failed enrollment is to
+		// look for "WebSDK CertRequest" or "Fix any errors, and then click
+		// Retry." in the retrieve response. We only check this on the first
+		// retrieve because we know that these two messages always disappear
+		// after resetting.
+		case statusCode == http.StatusInternalServerError && retrieveCount == 1 &&
+			(resp.Status == "This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry." ||
+				resp.Status == "WebSDK CertRequest Module Requested Certificate"):
```

Comment:
(this counter would have to be `==0` for a retry to happen)

Reply:
For context, Tim was getting that same message when one of Windows services was down, but that certificate wasn't in error.

@tr1ck3r In https://github.com/Venafi/vcert/issues/239#issuecomment-1242109588, you mentioned that TPP "will" return that body for a certificate object that was in error:

> It will have the following body after making an API request for a certificate object that was in error.
>
> ```json
> {
>   "Stage": 500, 
>   "Status": "WebSDK CertRequest Module Requested Certificate"
> }
> ```

Does this come from your knowledge of the TPP codebase, or is this something that you have observed over time?

Reply:
**tl;dr:** I'll go with the behavior we discussed, since "WebSDK CertRequest" means that there is a failed past enrollment in "sane" TPP instances.

**Context:** We had some back and forth with Tim. Since he has seen a case where TPP 22.04 would return the message "WebSDK CertRequest Module Requested Certificate" even though no prior enrollment existed, I wanted to investigate. Tim mentioned some Windows service was down that was causing the message. But I was not able to reproduce this issue. I assume it is a case of "misbehaving TPP" and won't happen in production TPP instances.

Tim suggested that we skip the "WebSDK CertRequest" messages until we find a "Click Retry" message.

I did a series of 100 tests with a TPP 20.1 server. Each time, I make sure the certificate is failing, then I request a certificate, and then I retrieve.

Case 1 occurs 60% of the time:

```
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
```

Case 2 occurs 30% of the time:

```
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
```

Case 3 occurs 10% of the time:

```
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
```

Results:
- We cannot only solely on "Click Retry": we would miss 10% of the cases that we are trying to fix (cf. case 3).
- We cannot, in my mind, skip the "WebSDK CertRequest" messages until we see the "Click Retry" message, because it means we would miss 10% of the cases that we were trying to fix. Another minor problem with this approach is that we would end up waiting until the end of the timeout, and not reset.